### PR TITLE
GDN recurrent scan kernel: added async dma for prefill, decode, and transition

### PR DESCRIPTION
**Pull request by yuhangning815**
# Description
image

This pull request optimizes the performance of the Gate Delta Net recurrent scan kernel by introducing asynchronous Direct Memory Access (DMA) for memory operations during the prefill, decode, and transition phases. For decode, the kernel now uses a double-buffering strategy (using slot 0 and slot 1) for loading and storing states. 

However, when combined with attention DP (which drops chunk_size 64->32, and increase the number of v heads 8->64), regular/transition prefill regressed, and the E2E throughput drops by 2% compared to baseline. This is because the CL added ~+8 MB of explicit decode-side scratch buffers in VMEM in this mem bound scenario of attention DP. 

Since the decode async buffers cannot be removed, the strategy is to reduce live transients and redundant VMEM accesses in regular prefill and transition prefill. All changes preserve bit-exactness.

The further optimization I did was:
- Alias state_commit_scratch to decode_store_scratch[0] since prefill and decode happens in serial.
- Transition prefill: per-iter loop cleanup (avoiding a full-state zeros tensor materialization, clean up the unconditional castings) 
- Regular prefill: transpose-before-repeat and matmul fusion. 

## Decode Schedule:
start load(0) -> decode_load_scratch[0] # async 
start load(1) -> decode_load_scratch[1] # async

Per iter b: (1) wait load @ slot    # waits for current iter DMA from iter b-2 
(2) start load @ slot for iter b+2     # prefetch async
(3) ===== COMPUTE: q/k/v, gates, attn, h_new =====
(4) wait store @ slot from iter b-2
(5) decode_store_scratch[slot] = h_new
(6) start store @ slot for iter b -> HBM

drain slot 0 store
drain slot 1 store

## Prefill Schedule:
(1) start init load -> state_commit_scratch # async 
(2) ===== chunk-prep compute (~100 lines): ====
q/k/v extract, gates, g_cumsum, S, S_q, A_inv, u, w, q_g
(3) wait init load -> prefill_scratch[slot]

# GPU comparison
In GPU, it's different - all decodes are padded as prefill if there is a prefill req in the batch. No double buffer or async DMA needed for GPU since all requests are processed in parallel by different CTAs.

# Tests
Kernel benchmark:
decode=13, prefill=[8, 16, 24]: 118.3us -> 96.1us.

Evals: MMLU Pro 
| Groups |Version|    Filter    |n-shot|  Metric   |   |Value |   |Stderr|
|--------|------:|--------------|-----:|-----------|---|-----:|---|-----:|
|mmlu_pro|      2|custom-extract|     5|exact_match|↑  |0.8314|±  |0.0138|

Qwen3.5 E2E Benchmark:
8k/1k, conc 512:
Output token throughput (tok/s):         2,162 -> 2406.59 
Total Token throughput (tok/s):          19,491 -> 21696.41

1k/8k conv 512:
Output token throughput (tok/s):         5,647 -> 5676.48
Total Token throughput (tok/s):          6,352 -> 6385.02

### CODE REVIEW

**Comment by yuhangning815** [comment]
@coolkp

File: tpu_inference/kernels/gdn/recurrent_scan_v2.py [hunk 1]
```diff
@@ -176,24 +181,70 @@ def get_target_idx(b):
...
             @pl.when(is_valid)
             def do_work():
-                target_idx = get_target_idx(b)
-
-                # Load state TODO: make async
-                copy_op = pltpu.make_async_copy(
-                    src_ref=recurrent_state_in.at[pl.ds(target_idx, 1)],
-                    dst_ref=state_commit_scratch,
-                    sem=decode_read_semaphores.at[0],
+                # Wait for THIS iter's load (issued by preload or by iter b-2 prefetch).
```
Comment (coolkp) [inline tpu_inference/kernels/gdn/recurrent_scan_v2.py:252]: nit: "this"

File: tpu_inference/kernels/gdn/recurrent_scan_v2.py [hunk 2]
```diff
@@ -176,24 +181,70 @@ def get_target_idx(b):
...
-                decode_state_scratch[pl.ds(
-                    0, 1)] = state_commit_scratch[...].astype(jnp.float32)
+                wait_load.wait()
+
+                # Copy loaded state to fp32 compute scratch.
+                decode_state_scratch[pl.ds(0, 1)] = decode_load_scratch[pl.ds(
+                    slot, 1)][...].astype(jnp.float32)
+
+                # Prefetch load for iter b+2 (same slot, since (b+2) % 2 == b % 2).
+                # This DMA overlaps with the compute below.
```
Comment (coolkp) [inline tpu_inference/kernels/gdn/recurrent_scan_v2.py:267]: can you be specific which compute it is

File: tpu_inference/kernels/gdn/recurrent_scan_v2.py [hunk 3]
```diff
@@ -341,23 +378,68 @@ def do_work():
...
                 copy_op.start()
-                copy_op.wait()
-
-                return None
-
-            return None
+                # No wait — drained after fori_loop.
+
+            # Update carry: this iter marked its slot as having an in-flight
+            # store iff is_valid.
```
Comment (coolkp) [inline tpu_inference/kernels/gdn/recurrent_scan_v2.py:435]: nit: if

**Review by coolkp — APPROVED** [review]

**Comment by kyuyeunk** [comment]
Can you rebase correctly? There's a lot of unrelated code changes in this pr.

**Comment by yuhangning815** [comment]
@kyuyeunk Yes now it rebases correctly. 

I accidentally rebased it to other branches to do some testing (should've used a new branch for testing)

**Review by kyuyeunk — APPROVED** [review]

**Comment by yuhangning815** [comment]
with attentionDP, chunk size becomes 32 and number of v heads becomes 64 -> double buffering adds more VMEM usage and spills and leads to perf regression for prefill and transition prefill. Decode still has the perf benefit. 

The previous e2e numbers are before attention DP. 

I am trying to further optimize this kernel. Hang tight.

File: tpu_inference/kernels/gdn/v2/recurrent_scan_v2.py [hunk 4]
```diff
@@ -15,14 +15,44 @@
...
 from jax.experimental.pallas import tpu as pltpu
+import jax.numpy as jnp
 
-from tpu_inference.kernels.gdn.v2 import \
+from tpu_inference.kernels.gdn import \
     compute_schedule_v2 as compute_schedule_table_v2
 
 
+
+# This is actually slower in the fused kernel. But it's faster in unit tests. likely it's due to VPU pressure.
```
Comment (coolkp) [inline tpu_inference/kernels/gdn/v2/recurrent_scan_v2.py]: can you remove this block?

Comment (yuhangning815) [reply tpu_inference/kernels/gdn/v2/recurrent_scan_v2.py]: done

**Review by helloworld1 — APPROVED** [review]
